Omnibus LwM2M changes for v1.14 (including socket API migration)

mike-scott opened the pull request:
This is a WIP PR so that some key stake holders can discuss if this approach is feasible.

After attempting to convert the LwM2M subsys lib to socket API, I realized that many devices don't have the hardware resources necessary to run in a secure state.  Previously, this was possible under the net-app APIs.   To that end, I don't believe we can make a direct cut over and should support net-app APIs until a better solution is presented.

Goals:
- Keep support for net-app while adding support for socket API via Kconfig option (socket API could be the default if desired)

Discussion / Thought process:
- net_buf provides a well supported method for building protocol data that needs buffers which can be appended / inserted to.  They can be configured to use a data pool in a memory efficient way.
- net_pkt APIs have several functions which operate primarily on the buffer fragments contained in the packet.  Several of these would be useful to have in the net_buf APIs, but a net_buf_context structure is needed for some of them (read as "insert) to hold information such as a buffer allocator function, etc. (NOTE: I'm currently using "appendable_net_buf" as the name of the context structure... open to ideas here.)
- Once we have the needed net_buf APIs supported (insert, etc), we can drop net_pkt specific behavior in the CoAP lib and the backend formatter functions in the LwM2M lib and instead only use net_buf_context(s).
- For a subsys like LwM2M where the actual connection is managed, we can add a network layer API abstraction, so that both socket and net-app can live together in harmony.
- Other subsys like MQTT and HTTP could also use this method.
- Later when https://github.com/zephyrproject-rtos/zephyr/issues/7590 (sockets to use net_bufs) the protocols will already be in a state to support it.

WHAT WORKS:
- Select LwM2M network layer via the following Kconfig:
LWM2M_NET_LAYER_NET_APP=y
or
CONFIG_LWM2M_NET_LAYER_SOCKET=y
- Tested samples/net/lwm2m on qemu_x86 target and swapped between net-app and socket layers using the overlay-samples.conf.

Here are some size comparisons of the LwM2M client sample built for BLENano2 (nRF52832)
```
[NET-APP]:
Current master branch (w/o PR):
Memory region         Used Size  Region Size  %age Used
           FLASH:      198060 B       512 KB     37.78%
            SRAM:       43908 B        64 KB     67.00%
        IDT_LIST:         136 B         2 KB      6.64%

NEW (w/ PR):
Memory region         Used Size  Region Size  %age Used
           FLASH:      197960 B       512 KB     37.76%
            SRAM:       44228 B        64 KB     67.49%
        IDT_LIST:         136 B         2 KB      6.64%

[SOCKETS]
Memory region         Used Size  Region Size  %age Used
           FLASH:      206092 B       512 KB     39.31%
            SRAM:       50436 B        64 KB     76.96%
        IDT_LIST:         136 B         2 KB      6.64%
```

TODO:
- Add socket DNS support
- Add socket DTLS support
- Split this up into normal patches for actual submission
- Simplify tests/net/lib/coap by removing net_pkt usage and IP headers

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
@nashif @jukkar @tbursztyka @laperie @pfalcon  Please let me know if this is a direction we can go for socket API support.  If so, I'll continue to break this down into acceptable upstream patches.

mike-scott commented:
Updated all the samples / tests I could find using CoAP APIs.  Let's check sanity now.

rveerama1 commented:
@mike-scott I am also working on CoAP over sockets support from past couple of days. I did't know that you went very far. My approach is quite different than yours. I just modified only net_pkt related stuff from CoAP library. But your approach is changing a lot in net_buf, which is not required IMO. My changes are at initial stage. I will create WIP/RFC PR by EOD.

rlubos commented:
I've briefly glanced through the changes, and without diving into details I have one general remark/question. Is it still a good approach to use `net_buf`/`net_pkt` to create CoAP/LWM2M messages as we aim for socket API now? For sockets, it brings additional overhead, as we need to lineraze the data, just to have it packed into `net_buf`s again at socket layer.

pfalcon commented:
> Is it still a good approach to use net_buf/net_pkt to create CoAP/LWM2M messages as we aim for socket API now?

It's definitely not. A litmus test for doing it right would be that the development actually happens on Linux, as more comfortable, and then everything just works on both Zephyr and Linux. (That's how for example https://github.com/zephyrproject-rtos/zephyr/pull/5985 was done.)

rveerama1 commented:
@mike-scott This is my branch (https://github.com/rveerama1/zephyr/commits/coap_sock) and commit is https://github.com/rveerama1/zephyr/commit/9f0647e63cedb5dd15fe9e2c4cbda8e764e5cc46. This is exactly same as current CoAP without net_pkt stuff in CoAP library. As I said just started few days back, it's work in progress. coap-client sampel don't even compile :). Just for reference about my approach.

mike-scott commented:
@rveerama1 Your method seems to add a 2nd CoAP library for socket support?  Are we sure we want to maintain and test 2 libs?  This seems prone to bit rot.

jukkar commented:
> Are we sure we want to maintain and test 2 libs?

I think the plan is to deprecate the older net_buf based one. The grand master plan is to provide only socket based interfaces to applications. Internally net_buf's will be used later too.

pfalcon commented:
> Internally net_buf's will be used later too.

But those may be reasonably different net_buf's after https://github.com/zephyrproject-rtos/zephyr/issues/7578. Churn is everywhere! Sockets are the only safe heaven ;-).

mike-scott commented:
@rveerama1 : Did you compare the resource usage before / after your PR?

@rlubos 
> I've briefly glanced through the changes, and without diving into details I have one general remark/question. Is it still a good approach to use net_buf/net_pkt to create CoAP/LWM2M messages as we aim for socket API now? For sockets, it brings additional overhead, as we need to lineraze the data, just to have it packed into net_bufs again at socket layer.

To clarify: The PR moves ``net_pkt`` usage into net-app only code.
So we're discussing ``net_buf`` usage.  One could say that using flat buffers will actually require even more resources than using ``net_buf``.  Also, ``net_buf`` isn't going anywhere.  It will be used in the future as it supports all the nice read / write / append / insert APIs in a nice dynamic way.

@pfalcon 
> It's definitely not. A litmus test for doing it right would be that the development actually happens on Linux, as more comfortable, and then everything just works on both Zephyr and Linux. (That's how for example #5985 was done.)

I think the keyword you said is "comfortable" and not "doing it right".  The reality is: it won't be like Linux programming.  Zephyr is an RTOS aimed at small HW.  Have you seen how careless Linux programmers are with resources?  While it might provide a general direction, your statement is filled with unicorns and fairy tales.

There is a very concerning trend of heavier resource usage currently happening in Zephyr.  I hope we keep in mind that **currently** you can run a bluetooth stack, IP-stack, and an LwM2M client using DTLS all on HW with 64K RAM and 512K flash in a way where you can have room for 2 copies of the application in order to perform OTA updates in a secure manner.

Very quickly we wont:
- Sockets currently require flat buffers that are hard to estimate in size using a protocol like LwM2M
- Sockets currently require NewLibC
- Logger is being introduced which currently has an unknown effect on RAM / flash usage
- Userspace support adds additional overhead

There has to be a way that we can still support the "simple" RTOS use-cases.

I illustrated the resource jump in my PR description and I chose a fairly conservative approach for sharing multi-use ``net_buf`` behind the scenes (many packets for instance are often less than 128 bytes).   If I was to only using max MTU flat buffers the RAM usage would be even higher.  Ignoring the fact that we currently have no read / write / append / insert APIs for flat buffers which would almost certainly need to be developed.

If you look at companies introducing products into the market, they are already looking for ways to cut cost.  The last thing they are going to do is "buy bigger" so they can use Zephyr.  They'll just go elsewhere.

@jukkar 
> I think the plan is to deprecate the older net_buf based one. The grand master plan is to provide only socket based interfaces to applications. Internally net_buf's will be used later too.

I think you meant to say net-app APIs are being depreciated, and that ``net_buf`` is being optimized/redesigned.   This PR moves away from being `net_pkt` dependent and stays with ``net_buf``s which will continue to be used in the future.

There is confusion between "applications" and "internal network library code" atm that probably needs to be ironed out.

@pfalcon 
> But those may be reasonably different net_buf's after #7578. Churn is everywhere! Sockets are the only safe heaven ;-).

Not so fast: https://github.com/zephyrproject-rtos/zephyr/issues/7590

mike-scott commented:
@rveerama1 
I'm not sure your approach is going to work well with:
```
/* CoAP Message */
static u8_t message[MAX_COAP_MSG_LEN];
static u8_t response[MAX_COAP_MSG_LEN];
```

This means the CoAP APIs are not thread safe?

rveerama1 commented:
> @rveerama1 : Did you compare the RAM / flash resource usage before / after your PR?

Like I mentioned already, sample doesn't even compile.

jukkar commented:
> This means the CoAP APIs are not thread safe?

Note that those static buffers are only in the sample application, the library itself is just using what ever buffer that is given to it.

pfalcon commented on include/net/buf.h:
```diff
@@ -1332,29 +1332,170 @@ int net_buf_linearize(void *dst, size_t dst_len,
+/**
+ * @brief Get a byte value from fragmented buffer
+ *
+ * @param frag Network buffer fragment.
```
Copy-paste errors/confusion (?), missing opportunity to improve the API. E.g.: 

>offset Offset of input buffer

which "input buffer" and "offset **of**"??


> pos Pointer to position of offset

Position of offset? What's that, in plain words?

Why these are still 2 params, if at least 95% of the code has those 2 params as `pos, &pos`?

> after reading 2 bytes

Not 2, 1 here.

> value Value is returned

Value is returned?

pfalcon commented:
@mike-scott re: https://github.com/zephyrproject-rtos/zephyr/pull/9832#issuecomment-419521236

I don't know where to start ;-). Let go quick over some statements.

>  Have you seen how careless Linux programmers are with resources?

Yup ;-). But that doesn't mean that POSIX API is resource-bloat - it's carefully thought out and designed to balance well overheads, usability, and underlying technology abstraction.

> There is a very concerning trend of heavier resource usage currently happening in Zephyr.

We can even have a case study why ;-). So, some time ago one programmer decided to add a framework for "modem drivers". It went as low-level as handling IRQs from UART. He was pointed that there's already an API in Zephyr which is very close to his needs, and abstracts from the need to deal with UART IRQ, but he decided to duplicate functionality anyway. So, any wonder that Zephyr code size bloats up, and RAM of course too, because nobody works on optimization, everyone just adds their "original designs".

P.S. Just had a look at your code again and paid attention that you put to a k_pipe from ISR. But see a Note at the end of http://docs.zephyrproject.org/latest/kernel/data_passing/pipes.html#concepts section.

> on HW with 64K RAM and 512K flash

That's too bloated IMHO, we should target for much smaller ;-).

> Sockets currently require flat buffers.

Socket API forever requires flat buffers. It's one of the biggest features. (No, it's not - it's just the obvious, baseline requirement.)

> LwM2M and other protocols may have a hard time estimating buffer sizes which present a growth problem.

Really? I thought that LwM2M is UDP based, so you can't send more than MTU at once.

> Sockets currently require NewLibC

Let's fix it if you're concerned.

> Logger is being introduced which currently has an unknown effect on RAM / flash usage
> Userspace support adds additional overhead

I'm with you on these ;-)

> I illustrated the resource jump in my PR description and I chose a fairly conservative approach for sharing multi-use net_buf behind the scenes

I see it a bit differently - as if you can't let those net_buf's go, and thus it precludes you to start from clean page and implement new exciting optimizations ;-).

Just as examples:

> (many packets for instance are often less than 128 bytes). If I was to only using max MTU flat buffers the RAM usage would be even higher.

Keep max MTU buffers only for LwM2M which can go to MTU. For less than 128 bytes messages, keep buffers not larger than 128 bytes.

> we currently have no read / write / append / insert APIs for flat buffers which would almost certainly need to be developed.

~~~
*ptr++ = val;
~~~

You can go up from there. Yeah, just imagine - you e.g. can make a message overflow check once, and if it passes, just write data directly, without any error checking. Compare that to calling 100 times  some function with gazillion of params, result of each call requiring a check. Do you see where bloat comes from?

> If you look at companies introducing products into the market, they are already looking for ways to cut cost. The last thing they are going to do is "buy bigger" so they can use Zephyr. They'll just go elsewhere.

Yeah, they go elsewhere, to RTOS which offer well-known, time-proven APIs. That's the way to cut cost - on R&D and support/maintenance. As for BOM, it doesn't have to be bigger. Again, compare direct pointer access vs calling some godawful functions with error checking.

pfalcon commented:
@mike-scott : And now some general comments.

It's a big achievement that we all came to conclusion that BSD Sockets API is a way forward in Zephyr. At the same time, there should be absolutely no worries about native (net_context-based) API and its wrappers like net_app. They aren't going anywhere until sockets-based solutions are proven to be at least not worse.

With that in mind, I would imagine that someone starting to work with sockets, does that out of excitement of using well-known API and simplifying the design that offers. However, with your code - and that's just my personal opinion - I don't see that approach. Instead I see an attempt to marry approach from old adhoc API with sockets. While one of the possibilities, I (again, personally), don't see a point in doing it like that - sockets allow to exercise new approaches, and old net_app doesn't go away. I may be wrong. After all, you know LwM2M and its architecture, in Zephyr and overall, so maybe it makes sense.

But note that trying to marry native API and sockets, you already introduce new entities like `struct appendable_net_buf`, so you're already doing a bunch of refactoring. But maybe refactoring for canonical sockets would make more sense?

Anyway, let me challenge myself in these conclusions in try to advocate your approach - in the next message.

pfalcon commented:
@mike-scott, So, summing up, I "reject" approach in this PR, in a sense that if you tell me that's the right approach, I don't believe you ;-). Note that I may be wrong.

But let me try to find independent, external references for such an approach which *could* have some similarity with approach used here.

Some time ago I stumbled upon this doc https://github.com/sustrik/bsd-socket-revamp/blob/master/source.txt from Martin Sustrik, a guy behind ZeroMQ and its rewrite nanomsg. The title is clickbait, of course it's not about "revamping" BSD Sockets API, it's perfect as it is. It's about designing new, "modern" networking API supposed to be of the same level as BSD Sockets API.

So, as you can see, he advocates usage of linked list of buffers for network I/O operations: https://github.com/sustrik/bsd-socket-revamp/blob/master/source.txt#L216 . Bingo!

Of course, there're big differences too. First of all, no talking about fixed-size buffers. So, @tbursztyka's refactor is what will be faithful implementation if this idea in Zephyr. Back to Sustrik's API, there's of course no "compact" function (which is hilarious one anyway), nor net_buf_read_u8 and friends with 4 params and return value which can signify error - such of course can be implemented, though many apps will get along without them. Nor there's net_buf_insert_bytes(), because presence of such a function signifies a design problem (protocol messages are constructed from beginning to end), but again, with linked lists it's possible and trivial.


So, much approach towards this? It's simple - we start with plain ol' good BSD Sockets and implement everything in their native manner. Then maybe we implement novelties like Sustrik's API - especially if he really submits it as an IETF RFC, etc. Implement on userspace level I mean. Because on kernel space we'll have it with @tbursztyka's refactor.

mike-scott commented:
@pfalcon 
Ok, so you're rejecting my proposal to re-use net_bufs as underlying dynamic memory structures in a single code base for CoAP and LwM2M which would work for both Sockets and Net-App.

First let me repeat some of my concerns:
1. Anything more than a zero-copy solution will cause a large amount of RAM increase in the networking layer for any given set of libraries (CoAP, LwM2M, HTTP, MQTT).   And it's worse that just *double * what a normal set of TX data and packets would be.  Each library will carve out their own set of max MTU flat buffers, and they aren't shared with any other libraries.
2. I can do the same as @rveerama1 and develop a 2nd fully stand-alone implementation of LwM2M using only flat buffers that works with Sockets.   And because this doesn't solve #1, we will now have 2 sets of code to maintain for an unknown amount of time (until #1 is actually solved).

> Some time ago I stumbled upon this doc <snip>

I like the idea of being able to pass something other than a flat buffer to network APIs.  It makes sense.   Even more so, if we're planning on a `net_buf` refactoring which ends up with a structure much like an iolist... why am I being asked to remove the concept of `net_buf` now?

I get the feeling that we're not really ready for a re-write of the subsystem protocols.

pfalcon commented:
@mike-scott: Off-topic, are you in "TCP rewrite" meeting right now? On IRC right now?

pfalcon commented:
@mike-scott 

> Ok, so you're rejecting

I'm "rejecting" (in quotes) in a sense described above.

>  single code base for CoAP and LwM2M which would work for both Sockets and Net-App.

Yes, because "my" requirements is to not work like net_app does, but work like sockets do. Those two work in very different ways, so marrying them is counter-productive in my outlook. You have different requirements and different outcomes.

> Anything more than a zero-copy solution will cause a large amount of RAM increase in the networking layer for any given set of libraries (CoAP, LwM2M, HTTP, MQTT)

That's hypothesis. Let's provide/disprove it, find problematic spots and fix them, get the bottom of it, then design "modern networking API with zero-copy, better than BSD sockets", a-la Sustrik (well, preferrably to avoid "original research" and use best practices, but you know, not all the time they exist).

>  Each library will carve out their own set of max MTU flat buffers, and they aren't shared with any other libraries.

Cool, that means that one library can't DoS another library ;-).

> I can do the same as @rveerama1 and develop a 2nd fully stand-alone implementation of LwM2M

You even have 3rd option - take existing LwM2M engine, there should be plenty on a market, play with it, and only write your own if it's really needed. But you'd say that they all are bad - they are intended to work with well-known APIs like sockets, sockets don't support zero-copy, so they're all doomed. It's a wonder how world survived without Zephyr's zero-copy-with-compact-wholes-in-buffers (which is oxymoron, you can't be "zero copy" and "compact") for so long ;-).

> why am I being asked to remove the concept of net_buf now?

Because you're trying to bring kernel-private data structure into userspace, with all its baggage of verbose funcs with weird params (which you consider a benefit). Nobody asking you to remove, I just express my personal opinion that it's not the best investment of effort - regarding cleanliness of the result (and that's among my requirements).

> I get the feeling that we're not really ready for a re-write of the subsystem protocols.

Apparently we aren't. But we're doing very important thing - searching for common ground and consensus on how it can, then should, be done.

pfalcon commented:
@mike-scott :

> I do as a matter of fact need the ability to insert bytes into whatever memory structure is used for building the LwM2M protocol payload.


> Spoiler alert: it's not a design flaw.

Let me try to do a spoiler alert too: a) either I'm about to learn something very, very new; or ... b) we, after detailed case analysis, will come to conclusion that in your particular usecase, it's indeed more practical to shuffle memory around, even though in general it's not a good practice.

Here's my move:

> Without literally doing everything twice

"Everything" what? Calculating the length of data before you start to send it? D'oh, definitely do that, even if you count it as "twice". Or you don't know the length of data before you start to send it? Then you should use appropriate protocol, which supports chunking. Example: HTTP's Transfer-Encoding: chunked or Websocket.

But here we talk about UDP-based protocol with max 1.5K message size anyway! How much is "twice" if we talk about such amounts of data?

And if you really don't want to do it "twice", and instead want to shuffle memories around (let me guess, you want to do it multiple times, like there can be multiples varlen-coded length fields, and you'll patch each one by one, patiently moving the tail again and again), you can. You'll just need to call memmove() with your bare hands, instead of hiding behind "advanced APIs" like net_pkt_insert(). All it takes is just embracing the flat buffers! ;-)

mike-scott commented:
> You even have 3rd option - take existing LwM2M engine, there should be plenty on a market, play with it, and only write your own if it's really needed. But you'd say that they all are bad - they are intended to work with well-known APIs like sockets, sockets don't support zero-copy, so they're all doomed. It's a wonder how world survived without Zephyr's zero-copy-with-compact-wholes-in-buffers (which is oxymoron, you can't be "zero copy" and "compact") for so long ;-).

@pfalcon You're assuming quite a lot here (and in a heavy-handed sarcastic way).  I did indeed do a fairly thorough search of what was available **and compatible** with the internal workings of the Zephyr 1.7/1.8 APIs.  I'm not really diabolically opposed to the evil socket API empire (note my use of quirky non-directional sarcasm that serves to only make light of my own statement), I'm just making sure we take into account some items I feel are important -- because I'm **willing to do the work**.

> Because you're trying to bring kernel-private data structure into userspace, with all its baggage of verbose funcs with weird params (which you consider a benefit). Nobody asking you to remove, I just express my personal opinion that it's not the best investment of effort - regarding cleanliness of the result (and that's among my requirements).

> (let me guess, you want to do it multiple times, like there can be multiples varlen-coded length fields, and you'll patch each one by one, patiently moving the tail again and again),

I'm going to stop you right here.  **The never-ending injection of unrelated commentary presenting your opinion as my own thought processes has to stop.**   Not because I care one way or the other, but because it's a distraction and reflects poorly on the Zephyr peer-review community.  There could be someone new to the Zephyr community reading this thread right now, wondering why they would subject themselves to such behavior.  Paul, you're a friendly guy in person and I know you're better than this.

If we stay on point without degenerating to "I bet you would use a bubble sort over a binary tree.."  I think we can actually come up with a compromise that not only solves this particular discussion, but will also provide an example for other subsystems needing the same work.

I can see value in cleaning up the net_pkt / net_buf APIs.  And that can be prioritized before adding socket support to the networking subsystems.   @jukkar and @tbursztyka might need to decide if that's the direction we need to take.  I realize that I don't know enough about upcoming networking re-work to make an informed decision at the moment.

pfalcon commented:
> The never-ending injection of unrelated commentary presenting your opinion as my own thought processes has to stop.

I must have a bad day today. Funnily, I didn't post any response to https://github.com/zephyrproject-rtos/zephyr/issues/9330, simply because if I did it so far, it would start with "Mike, this seems to be written with some idea in mind, and lists entailments of that idea, but I have no idea which *that idea* is. I actually like figuring what that idea might be, but that liking only makes me be always-aware that my figuring is guesswork, and some people even get confused by that. So I won't be doing any guesswork here." Too bad I got carried away and didn't follow that idea here.

That's also why asked if you'll be at Linaro Connect, because discussing such topics face to face is indeed easier.

> If we stay on point without degenerating to "I bet you would use a bubble sort over a binary tree.."

You just guessed me very well, I wanted to come to concrete things and show example of MQTT impl in Python based on "pre-calc the lengths" idea, salted by a reference to anecdote that sorting app written in Python can be faster than carelessly implemented one in C.

>  I think we can actually come up with a compromise that not only solves this particular discussion

So, it seems to me that we (and not even 2 of us, though I was writing directly to you, not some unfamiliar novice) approach the issue from so different angles and positions, that we need to scale a lot of ground and literally need to look for log in each other's eye to find a compromise. I made my move, thanks for not returning it and stopping me. I apologize.

mike-scott commented:
I'm going to attempt to summarize the above discussion so far:

Mike:
- Doesn't want to require NEWLIBC for socket use due resource requirements (means no malloc/free, etc. which many "off the shelf" LwM2M clients tend to use).  But, thinks the subsystem libraries can and should make use of added functionality **if** NEWLIBC is selected (float support, etc).
- Wants to keep a flexible buffer structure (currently `net_buf`) to hand off to shared libraries such as CoAP, LwM2M and MQTT.  This buffer structure can be flattened prior to use with sockets if needed or the sockets API can be adjusted to take it directly.  Otherwise, thread-based subsystem code that handles send/receive will need a pool of max MTU flat buffers to pull from when handling traffic.
- Wants the CoAP library to work with whatever networking layer is selected, not separate code for each supported network layer.
- Wants the LwM2M engine + data formatters to work with whatever networking layer is selected.  Not 2 code bases.
- Thinks the `net_pkt` and `net_buf` APIs are too cumbersome and could be cleaned up.  (This may be on someone's TODO list already).
- Agrees that LwM2M client needs a better implementation for handling large amounts of data in standard READ / DISCOVERY operations.  Currently, only certain use-cases use BLOCK-TRANSFER, the rest are a single large packet.  NOTE: This could potentially cause a lot of problems in resource constrained devices and needs to be thought about very carefully.  99% of all normal READ/WRITE ops don't overflow the standard 1280 byte MTU for IPv6.  And the few cases that normally do are handled for the most part.
- Doesn't want to implement something only to have to re-write it once more networking changes land.

Paul:
- In general: the more Zephyr code looks like standard POSIX-based code, the better.  Too much specialization in Zephyr at the moment.  This PR is another example of specialization.
- Would like to see the client-based subsystem code use userspace friendly memory structures like flat buffers.  If needed, Zephyr could add ease-of-use APIs for managing append / insert.
- Willing to look at a more modern network API like Sustrik's where linked list of buffers would be supported.
- Client-based subsystem code should have logic to determine when to use MAX MTU buffers (READ/WRITE?) vs. smaller more memory efficient buffers for handling small operations (UPDATE).  This should help with memory usage.
[Will fill in the rest with @pfalcon response]

More items for possible discussion:
- Could `net_buf` be cleaned up into something that works like Sustrik's `iolist`?
- Can the concept of a `net_pkt` be cleaned up, and all data manipulation APIs moved to `net_buf`(that's where the data is anyway right?)?  This way the primary piece of data that gets handed around would be `net_buf` and `net_pkt` could be an entirely internal networking object not shared outside of subsys/net/ip/.   Is this already happening?
- Not sure how flat buffers only used prior to send/recv socket calls can be optimized down to the same memory usage as the current `net_pkt`/`net_buf` "shared buffer" implementations.

pfalcon commented:
> Mike:
>    Doesn't want to require NEWLIBC for socket use

https://github.com/zephyrproject-rtos/zephyr/pull/9923 . Let that be an illustration that resolving issues with sockets may be easier than worshiping poor (in the sense "unhappy", "loveless", not "bad") code we *had to* write (note: not "your code") in Zephyr before.

mike-scott commented:
@pfalcon update: I'm going to let @rveerama1 get closer to a final PR for the CoAP library changes which use flat buffers for now.  During that time, I'll continue to refine my approach to the LwM2M library changes which will support sockets.  But nothing immediate.

mike-scott commented:
This is V3 of the socket support for LwM2M:
- Is a complete separate dir (lwm2m_sock) to be aligned with coap_sock and several others doing it this way
- Updated to use the new Logger subsys
- I performed quite a bit of clean up, and there's probably more to come.  Including reducing the heavy stack usage in the UDP receive functions by moving what were stack allocated variables into the lwm2m_message structure where they are allocated in a more controlled manner.  Also moved the nebulous "lwm2m_engine_context" contents into lwm2m_message as well.  In general, the lwm2m_message structure represents the lifetime of input/output structures needed for an LwM2M response right up to when the ACK is received back from a CON message.

A comparison of the BLENano2 (nRF52832) before/after:

net-app implementation from about 3 weeks ago (w/o Logger):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      198080 B       512 KB     37.78%
            SRAM:       43972 B        64 KB     67.10%
        IDT_LIST:         136 B         2 KB      6.64%
```

net-app implementation w/ the Logger changes:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      203848 B       512 KB     38.88%
            SRAM:       46308 B        64 KB     70.66%
        IDT_LIST:         136 B         2 KB      6.64%
```

socket-based implementation (now uses Logger as well):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      200492 B       512 KB     38.24%
            SRAM:       48004 B        64 KB     73.25%
        IDT_LIST:         136 B         2 KB      6.64%
```

mike-scott commented:
TODOs:
- DTLS support
- DNS support
- Probably merge the net_layer API abstraction back into the main lwm2m code now that this is a socket-only implementation
- Remove the stack usage for the periodic service logic (should save quite a bit of RAM)

mike-scott commented:
@rveerama1 This is using your CoAP patches for a base

mike-scott commented:
@jukkar Any thoughts on some options for reducing the flash/RAM usage for Logger that I can use for the lwm2m sample?

jukkar commented:
> Any thoughts on some options for reducing the flash/RAM usage for Logger that I can use for the lwm2m sample?

After looking the lwm2m_client sample app, you could do this:

**overlay-bt.conf**
* `CONFIG_NET_IPV4=n`
* `CONFIG_NET_CONFIG_NEED_IPV4=n`

**prj.conf**
* Do you need so many IPv4 addresses? Usually for IPv4 it is enough to say `CONFIG_NET_IF_UNICAST_IPV4_ADDR_COUNT=1`
* Not sure about lwm2m needs but you might be able to lower number of network contexts, try with `CONFIG_NET_MAX_CONTEXTS=2` for example.

For generic logging options, you could tweak these
* `CONFIG_LOG_BUFFER_SIZE` which is by default 1024 bytes long
* Especially if you enable `CONFIG_LOG_INPLACE_PROCESS`, then the `CONFIG_LOG_BUFFER_SIZE` can be lowered, see `subsys/logging/Kconfig` for details. This simulates the way the old logging worked, so basically one can choose between memory consumption vs. performance.

mike-scott commented:
@dbkinder sorry, I should have specified that this includes the first 4 patches of @rveerama1 work on CoAP for sockets here: https://github.com/zephyrproject-rtos/zephyr/pull/10121

Those patches will be dropped here when they are merged.

mike-scott commented:
Now that socket-friendly CoAP library has been added to Zephyr, I rebased this work to implement a socket-based LwM2M library.

There are still some discussion points:
- I have a buffer utility library as an include file in the first patch.  This should in theory be landed somewhere and not included in a .h.  Also, the CoAP library can benefit from shared code and a separate patch to re-integrate it into that code base should be looked at.
- There's the discussion of the library managing the sockets vs. the sample opening the sockets and managing them.  @pfalcon has expressed his dislike for the thread in the library which does that.

pfalcon commented:
> There's the discussion of the library managing the sockets vs. the sample opening the sockets and managing them. @pfalcon has expressed his dislike for the thread in the library which does that.

Well, now that's a bit hard to parse. The thread that you mention - is that a thread of discussion or a thread of execution? It couldn't be first, I couldn't express dislike for the discussion, I explicitly express dislike for the opposite - lack of discussion ;-). And for the thread-of-execution, the only thing I remembering is expressing *concern* that minimum memory usage and multiple threads don't mix well (due to the fact that each thread requires quite a no-nonsense amount of stack). Orthogonal to that, I expressed preference for following [Dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) pattern - that's the idea of letting users instantiate particular objects, and let "library" code just take references to them to work with. Using that pattern allows for (much) more flexibility in how something is done and what kind of applications can be developed with such a library.

pfalcon commented:
> I have a buffer utility library as an include file in the first patch. This should in theory be landed somewhere and not included in a .h. Also, the CoAP library can benefit from shared code and a separate patch to re-integrate it into that code base should be looked at.

Well, that's something I wanted to bring in the final review of https://github.com/zephyrproject-rtos/zephyr/pull/10125. As a usual disclaimer, there's absolutely no finger-pointing, just a plain description of the situation we have (from my PoV of course, but heck I spent a lot of time exactly trying to stay in loop of different people do - and bring different people in this loop). So, how it is from my PoV:

1. You, @mike-scott, started the discussion of the need for buffer handling helpers for sockets.
2. Soon after that, @rlubos posted https://github.com/zephyrproject-rtos/zephyr/pull/10125 . Of course, it had buffer helper routines - everyone needs that! As a continuation of our discussion, I pinged  @mike-scott with an example of what was, IMHO, a sub-ideal design of such buffer helper routines. There was no reply from @mike-scott.
3. But @rlubos showed that the world is a better place than it seems to be - he actually rewrote that code, even though I didn't explicitly request that. That definitely contributed to extended time it took to finalize MQTT lib.
4. So, when I did final review of that (yesterday!) I even wanted to joke along the lines of "initiative is punishable, @rlubos, you now would need to factor out that buffer handling code as a generic module for others to reuse". I didn't post that because: a) I'm not sure whether it qualifies as "kudos!" in the situation we have, when reviews take months; b) because nobody seems to care, in particular @mike-scott who started it all, now seems to be too busy and not interested in it.
5. Boy I was wrong with that, because the very next day @mike-scott came out with factored out module for buffer handling routines.

But what's inside?

> static inline int buf_insert(u8_t *dst, u16_t *dst_len, u16_t dst_size,
			     u16_t offset, u8_t *src, u16_t src_len)

Really? Well, that's not what was discussed in @rlubos' PR and what he implemented, apparently because he thought it makes sense. So, looks like the above code is nothing but LwM2M's private buffer helpers, just the same as MQTT has its own private helpers. Designing a generic reusable buffer handling library cannot be done in vacuum, the initial step of it is opening a dedicating ticket, inviting all interested parties, enumerating different ways to implement them, discussing pros and cons of each, and selecting the best compromise.

mike-scott commented:
Changes:
- Rebased on today's master
- Moved buf_util.h into lwm2m_sock lib dir
  (We will need to address all of the buffer handling going on in Zephyr in a future PR)

TODOs:
- Fix DTLS

mike-scott commented:
@nashif I noticed that the CoAP socket support was eventually smashed into the same lib dir as net_pkt.   Will that be a requirement here?

jukkar commented:
> TODOs:
> 
>     * Fix DTLS

Hi Mike, the deadline for 1.14 features is approaching fast. How much work there is still to get the lwm2m into a shape where we could remove the legacy net-pkt based CoAP?
Is there anything I could do to help you here?

mike-scott commented:
> > TODOs:
> > ```
> > * Fix DTLS
> > ```
> 
> Hi Mike, the deadline for 1.14 features is approaching fast. How much work there is still to get the lwm2m into a shape where we could remove the legacy net-pkt based CoAP?
> Is there anything I could do to help you here?

I fixed DTLS over the weekend.  I need to test DNS support.

But the largest issue is how to proceed?  Keep this PR as a separate dir?  Or re-do it on top of the existing subsys/net/lib/lwm2m and migrate it to sockets (temporarily keeping net_app support)?  I pinged @nashif but didn't get an answer.

jukkar commented:
>  Or re-do it on top of the existing subsys/net/lib/lwm2m and migrate it to sockets (temporarily keeping net_app support)

Just do it on top of existing lwm2m so we can remove the legacy coap and after that the net-app api. No need to keep the net_app version of lwm2m.

mike-scott committed:
net: lwm2m: create DTLS config layer for LwM2M

Currently, this will select the needed configs for LwM2M and net_pkt.
During the migration to socket APIs, the net_pkt selections will change
to socket-based selects.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: move to flat buffers

As part of the migration from net_app APIs to socket APIs, let's
stop referencing the net_pkt fragments throughout the LwM2M library.

Establish a msg_data flat buffer inside lwm2m_message and use that
instead.

NOTE: As a part of this change we remove the COAP_NET_PKT setting.
The COAP library reverts to COAP_SOCK behavior.

This doesn't mean we use sockets in LwM2M (yet), it only means we
use the socket-compatible COAP library which parses flat buffers
instead of net_pkt fragments.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: save remote address during setup

net_app contexts save the remote address and we use this during
observe notifications and pending handling.  If we move to another
network layer such as sockets, then the remote address becomes
harder to reference.  Let's save it as a part of the client
context.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: introduce input formatter private data

The JSON formatter is currently not enabled for incoming WRITE
operations.  To update the code in the formatter and not litter
the input context with extra data, let's allow formatters to
store their own user data.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: add JSON formatter for WRITE operations

Update the parsing functions for JSON used by the JSON data
formatter and enable it in the LwM2M engine.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: update security obj with DTLS data

For bootstrap support, we need to store connection credentials
in the security object.  This way the client can start a connection
at index 0 and after bootstrapping, move to the next connection.

Let's add the needed fields and a config item to set the key length.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: use security data for connections

In order to support bootstrap mode, we need to store server data
in the security / server objects.  Once the connection to the
bootstrap server is made, it will clear these objects and add
new server connection data.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: support for LwM2M bootstrap

Now that the security data can be loaded into and used from the
security / server objects, we can add support for LwM2M bootstrap.

This is a mode where initially a connection can be made to a server
which can update several LwM2M (including security and server
data) and then trigger a "bootstrap complete".  Once this happens
the client will start it's connection process over but now with
the new information.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: replace periodic service thread with work queue

We can save some resources by removing the periodic service thread
and replacing it by queuing the services to the work queue.

Before (reel_board using BT + DTLS)
Memory region         Used Size  Region Size  %age Used
           FLASH:      289464 B         1 MB     27.61%
            SRAM:       75620 B       256 KB     28.85%
        IDT_LIST:         136 B         2 KB      6.64%

After
Memory region         Used Size  Region Size  %age Used
           FLASH:      289576 B         1 MB     27.62%
            SRAM:       74596 B       256 KB     28.46%
        IDT_LIST:         136 B         2 KB      6.64%

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: fix firmware status after bad download attempts

This commit resets the firmware status to IDLE after a bad
download attempt.  Previously, the firmware object would stay
in an odd state and any further attempts to download firmware
would return an error.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
net: lwm2m: migrate LwM2M library to BSD-sockets API

This commit removes the net_app layer from the LwM2M library and
replaces it with BSD-sockets APIs.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
samples: lwm2m: migrate to BSD socket APIs

The LwM2M library has moved from the network application library
APIs to BSD socket APIs.  Let's make the needed changes in the
LwM2M sample to follow those changes.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
Changes with this update:
- Based on close to current master
- Patches now lay on top of subsys/net/lib/lwm2m/ (not in a separate directory)
- Initial pass at breaking up the patches into logical commits -- this may get re-worked a bit as feedback hits

Having the patches on top of the existing code highlights the changes a lot better:
- Migrate to socket APIs (and related changes like moving to flat buffers)
- LwM2M bootstrap support (and related changes like adding space in the objects for DTLS data)
- Added JSON formatter support for WRITE/CREATE operations
- Some cleanup

Coverity was down while I was re-working the patches, so those issues still need to be addressed.

@jukkar Please review and provide feedback

NOTE: I'm experiencing DTLS issues again and still in the process of debugging.

mike-scott commented:
4 of the "fix it" type patches from this PR are now here:
https://github.com/zephyrproject-rtos/zephyr/pull/12846

pfalcon commented:
> ~~LwM2M Socket Support~~ Omnibus LwM2M changes for v1.14 (including socket API migration)

What's this "Omnibus" part of the title?

jukkar reviewed (approved):
Thanks Mike, there was a lot of changes and I probably missed something, but overall looks good.

mike-scott commented:
> > LwM2M Socket Support Omnibus LwM2M changes for v1.14 (including socket API migration)
> 
> What's this "Omnibus" part of the title?

Just a funny literary reference: "a volume containing several novels or other items previously published separately."

It's a big PR of all of the changes to "complete" LwM2M 1.0 spec + migration to sockets.

mike-scott commented:
Rebased on todays master branch and removed merged patches

jukkar commented:
ping @dbkinder please re-review the changes

pfalcon reviewed (approved):
Approving on the idea that this needs to get into 1.14 anyway.

mike-scott committed:
net: lwm2m: add socket-based DNS support

Previously, the net_app layer handled DNS support as a part of
network initialization.  With the move to BSD-socket APIs,
we need to add support for DNS to the LwM2M library.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
Updated the commit with the following:
- Quite a few bug-fixes centered around cleaning up the sockets / connections when things go wrong
- Added DNS_RESOLVER support which uses socket-based getaddrinfo call
- Cleaned up IPv6 host parsing (sample adds brackets around IPs with colons in them)
- Use the same URL parser logic for firmware downloads (deleted the duplicate code)
- Removed a bunch of dead fields from the RD client structure

If we need to unblock the other net_app removal PRs, this can be merged now and I'll continue to run tests and shake out bugs over the next 2 weeks.

mike-scott commented:
For reference: bootstrap support, new JSON write operation parsing and the buf utilities make up most of the code additions (+600ish lines)

mike-scott commented:
Pushed a small bugfix for proxy firmware download

mike-scott commented:
Re-pushed to fix a line over 80 char warning.
